scripts: check_init_priorities: ignore zephyr,cdc-acm-uart  [WAS: usb: udc_nrf: change init priority to PRE_KERNEL1]

fabiobaltieri opened the pull request:
The new stack zephyr,cdc-acm-uart driver has two separate init path, one
of which kicks in before the USB stack to start buffering log messages.

This generates a false positive in the build time priority checking,
adding a check to ignore that compatible entirely.

~~Change the priority of udc_nrf so that it gets initialized before cdc_acm_uart. etc etc~~

jfischer-no reviewed (changes requested):
There is direct dependency between UDC driver an a CDC ACM instance, I habe to look at it probable find another fix if any.

fabiobaltieri commented:
> There is direct dependency between UDC driver an a CDC ACM instance, I habe to look at it probable find another fix if any.

Cool, we can tweak the script for special cases if necessary though I'd rather avoid it if possible, feel free to suggest a better fix.

jfischer-no commented:
> Cool, we can tweak the script for special cases if necessary though I'd rather avoid it if possible, feel free to suggest a better fix.

What I want to do is to remove cdc_acm_uart nodes from the bus, because virtual UART interface is actually other side of the implementation.
```diff
diff --git a/dts/bindings/serial/zephyr,cdc-acm-uart.yaml b/dts/bindings/serial/zephyr,cdc-acm-uart.yaml
index 8e0a003097..18926e8643 100644
--- a/dts/bindings/serial/zephyr,cdc-acm-uart.yaml
+++ b/dts/bindings/serial/zephyr,cdc-acm-uart.yaml
@@ -7,8 +7,6 @@ compatible: "zephyr,cdc-acm-uart"
 
 include: uart-controller.yaml
 
-on-bus: usb
-
 properties:
   tx-fifo-size:
     type: int
diff --git a/samples/subsys/usb/cdc_acm/app.overlay b/samples/subsys/usb/cdc_acm/app.overlay
index c50d0fd9d6..367c6dc721 100644
--- a/samples/subsys/usb/cdc_acm/app.overlay
+++ b/samples/subsys/usb/cdc_acm/app.overlay
@@ -4,7 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-&zephyr_udc0 {
+/ {
        cdc_acm_uart0 {
                compatible = "zephyr,cdc-acm-uart";
        };
```
But it would require a lot of changes in samples/applications. PRE_KERNEL_1 is too close to the clock drivers and should probably be avoided. Not sure what to do.

fabiobaltieri commented:
> But it would require a lot of changes in samples/applications. PRE_KERNEL_1 is too close to the clock drivers and should probably be avoided. Not sure what to do.

How about moving `cdc_acm_uart` to POST_KERNEL instead?

jfischer-no commented:
> How about moving `usbd_cdc_acm` to POST_KERNEL instead?

From the experience so far, all serial drivers should be PER_KERNEL. Deviations cause problems.

fabiobaltieri commented:
> From the experience so far, all serial drivers should be PER_KERNEL. Deviations cause problems.

Ok then I'm back proposing to anticipate the level of the usb driver. About clock, if there's a dependency with the clock controller, then that should be checked as well for priority by the very same script I'm using to detect this. That's not the case for `udc_nrf` since there's no clock device, but it is should be the case for other controllers.

fabiobaltieri committed:
scripts: check_init_priorities: ignore zephyr,cdc-acm-uart

The new stack zephyr,cdc-acm-uart driver has two separate init path, one
of which kicks in before the USB stack to start buffering log messages.

This generates a false positive in the build time priority checking,
adding a check to ignore that compatible entirely.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri commented on scripts/build/check_init_priorities.py:
```diff
@@ -54,6 +54,14 @@
 
+# List of compatibles for nodes where we don't check the priority.
+_IGNORE_COMPATIBLES = frozenset([
+        # This is pre-initialized early to start processing data as soon as
+        # possible, there's a separate SYS_INIT entry running later in the boot
+        # sequence.
```
@jfischer-no I took a good look at the code and I think it makes sense to just not check this, at least for now, does this sounds like a right summary of how https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device_next/class/usbd_cdc_acm.c#L1128 works?

jfischer-no replied:
This is not quite right, and we will probably change or remove SYS_INIT(). One thing is when the virtual driver is initialized, another is when the connection to the host is established, which is always later or never. I would write "There is no direct dependency between the CDC ACM UART and the USB device controller, the logical connection is established after USB device support is enabled. "

fabiobaltieri replied:
Cool, I've reworded it with your description, thanks!
